GH-3696: DeserializationEx support for KafkaMS

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/3696

* Add an internal logic into `KafkaMessageSource` to react properly for the `ErrorHandlingDeserializer` configuration and re-throw `DeserializationException`

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-3696: DeserializationEx support for KafkaMS

Fixes https://github.com/spring-projects/spring-integration/issues/3696

* Add an internal logic into `KafkaMessageSource` to react properly for the
`ErrorHandlingDeserializer` configuration and re-throw `DeserializationException`

artembilan commented:
In general just copy/pasted the logic from the `KafkaMessageListenerContainer`.
Not sure what could be done better though...

Thanks

garyrussell commented on spring-integration-kafka/src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java:
```diff
@@ -257,11 +271,68 @@ public Collection<TopicPartition> getAssignedPartitions() {
+	}
+
+	@Nullable
+	private Object findDeserializerClass(Map<String, Object> props, Properties consumerOverrides, boolean isValue) {
```
I will move most of the duplicated code to `SK.ErrorHandlingUtils`; then you can use it from there.

artembilan replied:
Sounds good!

garyrussell replied:
https://github.com/spring-projects/spring-kafka/pull/2756

artembilan committed:
* Use `ErrorHandlingUtils` from `spring-kafka` to determine a `deserializer` from consumer properties

garyrussell commented on spring-integration-kafka/src/test/java/org/springframework/integration/kafka/inbound/MessageSourceIntegrationTests.java:
```diff
@@ -122,4 +138,66 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+	@Test
+	void deserializationErrorIsThrownFromSource() throws Exception {
+		Map<String, Object> consumerProps = KafkaTestUtils.consumerProps(brokers, "testErrorChannelSource", "false");
+		consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
```
FYI, this is now the default in KTU.

garyrussell commented on spring-integration-kafka/src/test/java/org/springframework/integration/kafka/inbound/MessageSourceIntegrationTests.java:
```diff
@@ -122,4 +138,66 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+		String testData = "test data";
+		template.send(TOPIC2, testData);
+
+		source.receive(); // Trigger Kafka Consumer creation and poll()
```
I am not entirely sure, but the record might be returned by this initial poll.

It would be safer to remove this and move the following assertion into the `untilAsserted() -> {...}` block.

artembilan committed:
Revise the test in favor of just single `await().untilAsserted()` -
no reason to wait for assignment or check for returned record.
We just need to be sure that `DeserializationException` is thrown eventually
